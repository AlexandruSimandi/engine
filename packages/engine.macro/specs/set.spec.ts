import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

const macroFile = "'@c11/engine.macro'";

pluginTester({
  plugin,
  babelOptions: { filename: __filename },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should support Set": {
      code: `
        import { producer } from ${macroFile}
        const result = producer((
          a1 = Set.foo,
          a2 = Set[Prop.foo],
          a3 = Set[Arg.a2],
          a4 = Set.foo[Arg.bar.baz],
          a5 = Set.foo.bar[Arg.a2],
          a6 = {
            baz: Set.foo[Arg.a4].baz,
          },
          a7 = Set.foo[Arg.a5.baz],
          a8 = Set.foo[Param.prop],
          a9 = Set.foo[Param.prop].baz,
          a10 = Set.foo[Param.prop.a].baz[Prop.a2]
        ) => { })
      `,
      snapshot: true,
    },
  },
});
