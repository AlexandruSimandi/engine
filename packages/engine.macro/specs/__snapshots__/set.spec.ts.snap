// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros should support Set: should support Set 1`] = `

import { producer } from '@c11/engine.macro'
const result: producer = ({
  a1 = Set.foo,
  a2 = Set[Prop.foo],
  a3 = Set[Arg.a2],
  a4 = Set.foo[Arg.bar.baz],
  a5 = Set.foo.bar[Arg.a2],
  a6 = {
    baz: Set.foo[Arg.a4].baz,
  },
  a7 = Set.foo[Arg.a5.baz],
  a8 = Set.foo[Param.prop],
  a9 = Set.foo[Param.prop].baz,
  a10 = Set.foo[Param.prop.a].baz[Prop.a2]
}) => { }

      ↓ ↓ ↓ ↓ ↓ ↓

"use strict";

var result = {
  args: {
    type: "STRUCT",
    value: {
      a1: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
      a2: {
        type: "SET",
        path: [
          {
            type: "EXTERNAL",
            path: ["foo"],
          },
        ],
      },
      a3: {
        type: "SET",
        path: [
          {
            type: "INTERNAL",
            path: ["a2"],
          },
        ],
      },
      a4: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INTERNAL",
            path: ["bar", "baz"],
          },
        ],
      },
      a5: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: "bar",
          },
          {
            type: "INTERNAL",
            path: ["a2"],
          },
        ],
      },
      a6: {
        type: "STRUCT",
        value: {
          baz: {
            type: "SET",
            path: [
              {
                type: "CONST",
                value: "foo",
              },
              {
                type: "INTERNAL",
                path: ["a4"],
              },
              {
                type: "CONST",
                value: "baz",
              },
            ],
          },
        },
      },
      a7: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INTERNAL",
            path: ["a5", "baz"],
          },
        ],
      },
      a8: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INVOKE",
            path: ["prop"],
          },
        ],
      },
      a9: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INVOKE",
            path: ["prop"],
          },
          {
            type: "CONST",
            value: "baz",
          },
        ],
      },
      a10: {
        type: "SET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INVOKE",
            path: ["prop", "a"],
          },
          {
            type: "CONST",
            value: "baz",
          },
          {
            type: "EXTERNAL",
            path: ["a2"],
          },
        ],
      },
    },
    meta: {},
  },
  fn: function fn(_ref) {
    var a1 = _ref.a1,
      a2 = _ref.a2,
      a3 = _ref.a3,
      a4 = _ref.a4,
      a5 = _ref.a5,
      a6 = _ref.a6,
      a7 = _ref.a7,
      a8 = _ref.a8,
      a9 = _ref.a9,
      a10 = _ref.a10;
  },
};


`;
